weeks/week03/file-handling.md

```diff
@@ -0,0 +1,270 @@
+
+# Closing the file
+file.close()
+```
+
+
+## 1.2 Reading Files
+
+You can read the entire file, line by line, or only specific parts of the file.
+
+**Note:** Always use the `with` statement (context manager) to manage files. It ensures that files are properly closed after their operations, even if an exception is raised.
+
+
+### Example:
+
+```python
+'''
+The `read()` method reads the **entire file** content as a single string. You
+can optionally pass a number to specify how many characters to read. It's often
+used when you need to load everything at once, such as when dealing with smaller
+files.
+'''
+with open("test_data.txt", "r") as file:
+    content = file.read()
+    print(content)
+
+
+with open("test_data.txt", "r") as f:
+    partial_content = f.read(100)  # Reads the first 100 characters
+    print(partial_content)
+
+
+'''
+Read line by line. When dealing with large files, it's more efficient to read
+files line by line instead of loading the entire file into memory.
+'''
+with open("test_data.txt", "r") as file:
+    for line in file:
+        print(line)
+```
+
+
+### Using `readline()`
+
+The `readline()` method reads one line at a time from the file. This is useful when you want to process the file line by line, especially when the file is too large to fit in memory all at once.
+
+```python
+with open("test_data.txt", "r") as f:
+    first_line = f.readline()  # Reads the first line
+    print(first_line)
+    second_line = f.readline()  # Reads the second line
+    print(second_line)
+```
+
+
+### Using `readlines()`
+
+The `readlines()` method reads all lines in the file and returns them as a `list of strings`, where each element is a line. This method is useful when you want to read the entire file but still want to handle it line by line. Each line in the list will include the newline character (`\n`).
+
+```python
+with open("test_data.txt", "r") as f:
+    lines = f.readlines()  # Reads all lines into a list
+    for l in lines:
+        print(l)
+```
+
+
+## 1.3 Writing and Appending to Files
+
+You can `write` or `append` data to a file using `write()` or `append()` methods.
+
+````python
+# Writing to a file (overwrites if exists)
+with open("my_test_data.txt", "w") as file:
+    file.write("This is a test.\n")
+
+# Appending to a file
+with open("my_test_data.txt", "a") as file:
+    file.write("Appending this line.")
+````
+
+
+## 1.4 File Handling Exceptions
+
+Handling exceptions is crucial for error-prone operations like file handling.
+
+When using the context manager `with` in file handling, the context manager itself handles the automatic opening and closing of the file, even if an exception occurs during the operation. However, it does not handle exceptions related to file operations (like missing files or permission errors). Therefore, using a `try-except` block is necessary if you want to catch and handle exceptions such as `FileNotFoundError`, `PermissionError`, or other file-related errors.
```

Comment:
imho #1.4 should be combined with ## 1.9 Handling File Exceptions Gracefully in one place (e.g. for mention all FileHandling related exception)

Reply:
Done!